pkg/config/environment_detection.go

```diff
@@ -7,198 +7,98 @@ package config
 
 import (
 	"os"
-	"path"
-	"runtime"
 	"strconv"
 	"strings"
+	"sync"
 
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
 // Feature represents a feature of current environment
-type Feature int
+type Feature string
 
 const (
-	// Docker socket present
-	Docker Feature = iota
-	// Containerd socket present
-	Containerd
-	// Cri is any cri socket present
-	Cri
-	// Kubernetes environment
-	Kubernetes
-	// ECSFargate environment
-	ECSFargate
-	// EKSFargate environment
-	EKSFargate
-	// KubeOrchestratorExplorer can be enabled
-	KubeOrchestratorExplorer
-)
-
-const (
-	autoconfEnvironmentVariable = "AUTCONFIG_FROM_ENVIRONMENT"
-
-	defaultLinuxDockerSocket       = "/var/run/docker.sock"
-	defaultWindowsDockerSocketPath = "//./pipe/docker_engine"
-	defaultLinuxContainerdSocket   = "/var/run/containerd/containerd.sock"
-	defaultLinuxCrioSocket         = "/var/run/crio/crio.sock"
-	defaultHostMountPrefix         = "/host"
-	unixSocketPrefix               = "unix://"
-	winNamedPipePrefix             = "npipe://"
+	autoconfEnvironmentVariable         = "AUTOCONFIG_FROM_ENVIRONMENT"
+	autoconfEnvironmentVariableWithTypo = "AUTCONFIG_FROM_ENVIRONMENT"
 )
 
 // FeatureMap represents all detected features
 type FeatureMap map[Feature]struct{}
 
+func (fm FeatureMap) String() string {
+	features := make([]string, 0, len(fm))
+	for f := range fm {
+		features = append(features, string(f))
+	}
+
+	return strings.Join(features, ",")
+}
+
 var (
-	detectedFeatures = make(FeatureMap)
+	detectedFeatures FeatureMap
+	featureLock      sync.RWMutex
 )
 
 // GetDetectedFeatures returns all detected features (detection only performed once)
 func GetDetectedFeatures() FeatureMap {
+	featureLock.RLock()
+	defer featureLock.RUnlock()
+
+	if detectedFeatures == nil {
+		log.Critical("Trying to access features before detection has run")
```

Comment:
Because there's nothing to do if it happens. I considered changing `log.Critical` for a `panic()`, as it should not be possible to call `GetDetectedFeatures` before the detection is done, maybe I should switch to `panic()`.

Reply:
Moved to `panic()`